File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode detectCycle(ListNode head) {
        if(head == null)return null;
        //如何照顾边边角角
        ListNode fast = getIntersect(head);
        if(fast == null)return null;
        ListNode slow = head;
        while(slow != fast){
            slow = slow.next;
            fast = fast.next;
        }
        return fast;

    }
    public ListNode getIntersect(ListNode head){
        ListNode slow = head;
        ListNode fast = head;
        while(fast != null && fast.next != null){
            slow = slow.next;
            fast = fast.next.next;
            if(slow == fast){
                return slow;
            }
        }
        return null;
    }
}

File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var detectCycle = function(head) {
    if (head == null)
        return null;

    var doubleNode = head;
    var singleNode = head;
    
    while (doubleNode.next != null && doubleNode.next.next != null) {
        doubleNode = doubleNode.next.next;
        singleNode = singleNode.next;
        if (singleNode == doubleNode)
            break;
    }
    
    if (doubleNode.next != null && doubleNode.next.next != null && singleNode == doubleNode) {
        singleNode = head;
        var i = 0;
        while (doubleNode != singleNode) {
            doubleNode = doubleNode.next;
            singleNode = singleNode.next;
            i++;
        }
        return singleNode;
    }else {
        return null;
    }
    
    
};